// react-style-tag/lib/constants.js
'use strict';

exports.__esModule = true;
exports.PREFIXER = exports.ONLY_TEXT_ERROR = exports.NO_BLOB_SUPPORT_ERROR = exports.DEFAULT_REACT_STYLE_TAG_GLOBAL_PROPERTIES = undefined;

var _autoprefixer = require('autoprefixer');

var _autoprefixer2 = _interopRequireDefault(_autoprefixer);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @constant {Object} REACT_STYLE_TAG_GLOBAL_PROPERTIES
 */
// external dependencies
var DEFAULT_REACT_STYLE_TAG_GLOBAL_PROPERTIES = exports.DEFAULT_REACT_STYLE_TAG_GLOBAL_PROPERTIES = {
  doNotPrefix: false,
  hasSourceMap: false,
  isMinified: false
};

/**
 * @constant {string} NO_BLOB_SUPPORT_ERROR
 * @default
 */
var NO_BLOB_SUPPORT_ERROR = exports.NO_BLOB_SUPPORT_ERROR = '\nTo support sourcemaps for react-style-tag you need Blob support, and the browser you are using does not currently\nsupport it. Please import the included polyfill at \'react-style-tag/blob-polyfill.js\'.\n'.replace(/\r?\n|\r/, '');

/**
 * @constant {string} ONLY_TEXT_ERROR
 * @default
 */
var ONLY_TEXT_ERROR = exports.ONLY_TEXT_ERROR = 'The only type of child that can be used in the <Style/> tag is text.';

/**
 * @constants {Object} PREFIXER
 */
var PREFIXER = exports.PREFIXER = (0, _postcss2.default)([(0, _autoprefixer2.default)({
  remove: false
})]);
